### PR TITLE
aws_lightsail_sd

### PR DESCRIPTION
**Pull request by n888**
Hello @roidelapluie & Team,

Some notes:

- discovery/ec2 has been moved to discovery/amazon and merged with lightsail code (wondering if the naming/capitalizations are fit)
- I did not see any pagination "Pages" API related function for `GetInstancesWithContext`, also `GetInstancesInput` does not support a Filter, so I have not included any logic there. I assume these are missing because the lightsail instance limits are low (20/region).

Thank you in advance for your time and reviewing.

### CODE REVIEW

**Comment by roidelapluie** [comment]
Thank you! I will review the PR soon.

**Review by roidelapluie — COMMENTED** [review]
I have taken a first look. What do you think about moving lightsail in its own go file? Now this is interleaving ec2 and lightsail which is not really nice. e.g. `aws/aws.go aws/ec2.go aws/lightsail.go `?


You have forgotten the documentation.

**Comment by roidelapluie** [comment]
I have renamed dockerswarm to moby so you mught want to rebase this PR. What do you think about `aws` as package name BTW?

**Comment by n888** [comment]
Docs added, also added missing RoleARN code and split out ec2.go/lightsail.go.
I think `aws` is the better package name, as it is shorter, and seems to be more widely referenced.

File: discovery/amazon/amazon.go [hunk 1]
```diff
@@ -298,3 +420,70 @@ func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
...
+
+	tg := &targetgroup.Group{
+		Source: d.cfg.Region,
+	}
+
+	input := &lightsail.GetInstancesInput{}
+
+	output, err := lightsailClient.GetInstancesWithContext(ctx, input)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && (awsErr.Code() == "AuthFailure" || awsErr.Code() == "UnauthorizedOperation") {
```
[the remark this answers is not in the document]
Comment (n888) [reply discovery/amazon/amazon.go]: Please excuse my (beginner) dev skills, can you help me understand what consts I should be looking for?

File: discovery/amazon/ec2.go [hunk 2]
```diff
@@ -61,24 +61,26 @@ const (
...
 )
 
-// DefaultSDConfig is the default EC2 SD configuration.
-var DefaultSDConfig = SDConfig{
-	Port:            80,
-	RefreshInterval: model.Duration(60 * time.Second),
-}
+var (
+	// EC2DefaultSDConfig is the default EC2 SD configuration.
+	EC2DefaultSDConfig = EC2SDConfig{
```
Comment (roidelapluie) [inline discovery/amazon/ec2.go:66]:
```suggestion
	DefaultEC2SDConfig = EC2SDConfig{
```

File: discovery/amazon/lightsail.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
+// Copyright 2015 The Prometheus Authors
```
Comment (roidelapluie) [inline discovery/amazon/lightsail.go:1]:
```suggestion
// Copyright 20121The Prometheus Authors
```

File: discovery/amazon/lightsail.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+	lightsailLabelIPv6Addresses       = lightsailLabel + "ipv6_addresses"
+	lightsailLabelPrivateIP           = lightsailLabel + "private_ip"
+	lightsailLabelPublicIP            = lightsailLabel + "public_ip"
+	lightsailLabelTag                 = lightsailLabel + "tag_"
+	lightsailLabelSeparator           = ","
+)
+
+var (
+	// LightsailDefaultSDConfig is the default Lightsail SD configuration.
+	LightsailDefaultSDConfig = LightsailSDConfig{
```
Comment (roidelapluie) [inline discovery/amazon/lightsail.go:58]:
```suggestion
	DefaultLightsailSDConfig = LightsailSDConfig{
```

File: discovery/amazon/lightsail.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+	tg := &targetgroup.Group{
+		Source: d.cfg.Region,
+	}
+
+	input := &lightsail.GetInstancesInput{}
+
+	output, err := lightsailClient.GetInstancesWithContext(ctx, input)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && (awsErr.Code() == "AuthFailure" || awsErr.Code() == "UnauthorizedOperation") {
```
Comment (roidelapluie) [inline discovery/amazon/lightsail.go:184]: Are there consts for this?

File: docs/configuration/configuration.md [hunk 6]
```diff
@@ -238,6 +238,10 @@ hetzner_sd_configs:
 kubernetes_sd_configs:
   [ - <kubernetes_sd_config> ... ]
 
+# List of Lightsail service discovery configurations.
+lightsail_sd_configs:
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:242]: This also needs to go to the alerting SD section

File: docs/configuration/configuration.md [hunk 7]
```diff
@@ -1341,6 +1345,53 @@ for a detailed example of configuring Prometheus for Kubernetes.
 You may wish to check out the 3rd party [Prometheus Operator](https://github.com/coreos/prometheus-operator),
 which automates the Prometheus setup on top of Kubernetes.
 
+### `<lightsail_sd_config>`
+
+Lightsail SD configurations allow retrieving scrape targets from AWS Lightsail
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:1350]: Can you add a link to lightsail?

File: docs/configuration/configuration.md [hunk 8]
```diff
@@ -1341,6 +1345,53 @@ for a detailed example of configuring Prometheus for Kubernetes.
...
+The following meta labels are available on targets during [relabeling](#relabel_config):
+
+* `__meta_lightsail_availability_zone`: the availability zone in which the instance is running
+* `__meta_lightsail_blueprint_id`: the Lightsail blueprint ID
+* `__meta_lightsail_bundle_id`: the Lightsail bundle ID
+* `__meta_lightsail_instance_name`: the name of the Lightsail instance
+* `__meta_lightsail_instance_state`: the state of the Lightsail instance
+* `__meta_lightsail_instance_support_code`: the support code of the Lightsail instance
+* `__meta_lightsail_ipv6_addresses`: comma separated list of IPv6 addresses assigned to the instance's network interfaces, if present
+* `__meta_lightsail_private_ip`: the private IP address of the instance, if present
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:1363]:
```suggestion
* `__meta_lightsail_private_ip`: the private IP address of the instance
```

We mandate that

File: docs/configuration/configuration.md [hunk 9]
```diff
@@ -1341,6 +1345,53 @@ for a detailed example of configuring Prometheus for Kubernetes.
...
+
+# AWS Role ARN, an alternative to using AWS API keys.
+[ role_arn: <string> ]
+
+# Refresh interval to re-read the instance list.
+[ refresh_interval: <duration> | default = 60s ]
+
+# The port to scrape metrics from. If using the public IP address, this must
+# instead be specified in the relabeling rule.
+[ port: <int> | default = 80 ]
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:1393]: You have forgotten to close the code block (with <code>```</code>)

File: discovery/amazon/lightsail.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+		return nil, errors.Wrap(err, "could not get instances")
+	}
+
+	for _, inst := range output.Instances {
+		if inst.PrivateIpAddress == nil {
+			continue
+		}
+
+		labels := model.LabelSet{
+			lightsailLabelInstanceSupportCode: model.LabelValue(*inst.SupportCode),
```
Comment (roidelapluie) [inline discovery/amazon/lightsail.go:196]: Can we set the other here? like IP,  blueprint etc..

**Comment by roidelapluie** [comment]
Thank you. You have used amzon as name, can you switch to AWS? You will also need to fix the DCO and run go mod tidy for CI.

File: discovery/amazon/lightsail.go (hunk 5 above)
Comment (roidelapluie) [reply discovery/amazon/lightsail.go:184]: I have checked and could not find that so that is OK

File: discovery/amazon/ec2.go (hunk 2 above)
Comment (n888) [reply discovery/amazon/ec2.go:66]: Updated to `DefaultEC2SDConfig` & `DefaultLightsailSDConfig`

File: discovery/amazon/lightsail.go (hunk 3 above)
Comment (n888) [reply discovery/amazon/lightsail.go:1]: Updated copyright in `ec2.go` & `lightsail.go`

File: discovery/amazon/lightsail.go (hunk 4 above)
Comment (n888) [reply discovery/amazon/lightsail.go:58]: Updated

File: docs/configuration/configuration.md (hunk 7 above)
Comment (n888) [reply docs/configuration/configuration.md]: https://github.com/prometheus/prometheus/pull/8693/commits/3db49731a807b670e91fc84ce6f402f19cb8a0df

File: docs/configuration/configuration.md (hunk 8 above)
Comment (n888) [reply docs/configuration/configuration.md]: https://github.com/prometheus/prometheus/pull/8693/commits/ffdb9a687b15ea33962887a9110e610986c8bd99

File: docs/configuration/configuration.md (hunk 9 above)
Comment (n888) [reply docs/configuration/configuration.md:1393]: https://github.com/prometheus/prometheus/pull/8693/commits/1981dc9c84393bb09e1ecffaf43b2dc10d4a817e

File: discovery/amazon/lightsail.go (hunk 10 above)
Comment (n888) [reply discovery/amazon/lightsail.go]: https://github.com/prometheus/prometheus/pull/8693/commits/45b8e9f303661f48eda638d9ad9fee65ee6aec95
Do we want any more labels inside here?

File: docs/configuration/configuration.md (hunk 6 above)
Comment (n888) [reply docs/configuration/configuration.md:242]: https://github.com/prometheus/prometheus/pull/8693/commits/ff78f97d5139e5c9f4b171c7536c9e3ec11d840c

**Comment by n888** [comment]
> Thank you. You have used amzon as name, can you switch to AWS? You will also need to fix the DCO and run go mod tidy for CI.

Much appreciated,

 - aws rename -> https://github.com/prometheus/prometheus/pull/8693/commits/73fe6450d2a2193fd95792e561f0ebc698f5ca2b

**Review by roidelapluie — APPROVED** [review]
Thank you, this is ready to be merged once you fix the tests (run go mod tidy I think)

**Comment by n888** [comment]
Thanks, so I got the go mod tidy sorted by upgrading to go 1.16.3. Any idea what's going on with `ci/circleci: test_windows` ?

**Comment by roidelapluie** [comment]
Thank you!!!!!

**Pull request merged**